tools/src/main/java/org/apache/kafka/tools/KafkaClientPerformance.java

```diff
@@ -126,28 +97,30 @@ public static void main(String[] args) throws Exception {
 
             int currentTransactionSize = 0;
             long transactionStartTime = 0;
+            long lastFlushTime = 0;
             for (long i = 0; i < numRecords; i++) {
                 if (transactionsEnabled && currentTransactionSize == 0) {
                     producer.beginTransaction();
                     transactionStartTime = System.currentTimeMillis();
                 }
 
-
-                if (payloadFilePath != null) {
-                    payload = payloadByteList.get(random.nextInt(payloadByteList.size()));
-                }
-                record = new ProducerRecord<>(topicName, payload);
+                record = clientTestHandler.getRecord();
```

Comment:
Since it looks like we expect numRecords to always be set, should we error out here if record is null? That would seem to indicate that our source topic was under-sized or some other issue (GC pause?) caused poll to timeout.

Reply:
I added a null check to be safe here.